### PR TITLE
Foreach intents first steps

### PR DESCRIPTION
**Pull request by stonea**
This PR begins work on enabling intents on `foreach` loops.  Specifically, it changes things so that this example now properly errors out (saying you can't assign to the constant 'c'):

```chpl
var c = 42; 
foreach i in 0..0 with (const c) do c = 1;
```

While ensuring that an example that doesn't modify 'c' still works fine. It also allows `ref` intents syntactically, so this will work:

```chpl
foreach i in 0..0 with (ref x) do x = 52;
```

Although at this point the `ref` intent itself won't change any existing behavior (so for example if this loop is gpuized and run on the gpu it will keep the existing behavior where `x` is passed to the kernel by-value effectively treating like an `in` intent).

Anyway, this is all a work in progress but I would like to get code merged early and often as I make progress on it.  With this in mind I introduce a new feature controlling global `fForeachIntents` in `driver.h` that can be turned on/off by passing `chpl` `--foreach-intents`.  For the time being I'm leavning this default "off".

When this feature is off foreach loops that have explicit intents will error with a syntax error early on (during `convertuast`) and at the moment we don't apply implicit intents to `foreach` loops that don't have any other explicit intents.

**Implementation Notes**:

- Before being lowered, `foreach` loops are modeled in the production compiler's AST using the `ForLoop` node. `forall` loops use the `ForallStmt`.
  - Unfortunately, the nearest common ancestor of these nodes is `Expr` so there isn't really a nice place to put common methods or data members.  I spent some time trying to see if could mock up a way to refactor things so they could be (see the "Exploring if we could rearrange the class hierarchy" comments here https://github.com/Cray/chapel-private/issues/5092#issuecomment-1648278779) and I don't think it's really feasible without major engineering efforts. To avoid duplicating code as much as I can I create a class `ShadowVarLoopInterface` that I have both `ForLoop` and `ForallStmt` derive from.
- There are several places where I either copy or reuse code to get `foreach` loops to behave similarly to forall loops:
  - **Convert uast**: if the user opts into `foreach` intents then I skip a check that would previously error out. I then pass along intents to a `ForLoop::buildForeachLoop` function and process it there similarly to what we do for ForAll loops.
  - **Scope resolve**: there's a loop that iterates over all forall loops and processes their shadow variables, it's just a matter of copying this loop and doing the same for foreach (thankfully the bulk of the work done by the loop is already factored out into functions that are written generically enough to work for both kinds of loops). 
  - **Function resolution**: during resolveExpr when we encounter a forall loop we call call `setupAndResolveShadowVars` in `implementForallIntents.cpp`. I've expanded this to also call `setupAndResolveShadowVars` for foreach loops when they have a shadow variable (which happens if the user explicitly used a `with` clause to add intents).
    - Eventually I'd like to modify things so this is always called (so that we set implicit shadow variables for loops the user didn't explicitly imply intents on) but I'm leaving this as future work.
    - I've modified the various functions in this file to operate on my new `ShadowVarLoopInterface` interface instead of on `ForallStmt` directly.
      - There is some code that I think really is `ForallStmt` specific, in which case I have an "escape hatch" in `ShadwoVarLoopInterface` to see if the loop really is a forall loop and if so get a pointer back to it (see uses of `isForallStmt`).

**Testing**:

I have run this through (non gasnet non gpu related) paratest and it seems to pass fine without breaking anything regardless of whether the feature is enabled or disabled.

### CODE REVIEW

File: compiler/AST/ForLoop.cpp [hunk 1]
```diff
@@ -635,3 +654,7 @@ Expr* ForLoop::getNextExpr(Expr* expr)
 
   return retval;
 }
+
+bool ForLoop::isInductionVar(Symbol* sym) {
+  return sym == mIndex->symbol();
+}
```
Comment (e-kayrakli) [inline compiler/AST/ForLoop.cpp:660]: I am guessing you were able to get by with this because:

```chpl
for (i,j) in My2DThing
```

`(i,j)` is packed in a tuple. That sound right?

(This is one of the things that really bothers me. I fixed it for foralls in the past, but I had to scrap my branch that did it for `forall`s)

File: compiler/AST/symbol.cpp [hunk 2]
```diff
@@ -1032,8 +1033,14 @@ void ShadowVarSymbol::verify() {
...
-    INT_ASSERT(pfs);
-    INT_ASSERT(defPoint->list == &(pfs->shadowVariables()));
+    ForLoop *pfl = toForLoop(defPoint->parentExpr);
+    if(pfs) {
+      INT_ASSERT(defPoint->list == &(pfs->shadowVariables()));
+    } else {
+      INT_ASSERT(pfl);
+      INT_ASSERT(pfl->isOrderIndependent());
+      INT_ASSERT(defPoint->list == &(pfl->shadowVariables()));
+    }
```
Comment (e-kayrakli) [inline compiler/AST/symbol.cpp:1043]: This part feels a bit redundant with `pfs` and `pfl`. Consider:

```cpp
if (ForallStmt* pfs = ...) {
  INT_ASSERT(... pfs ...);
}
else if (ForLoop* pfl = ...) {
  INT_ASSERT(... pfl ...);
}
else {
  INT_FATAL(defPoint, "intent on an unexpected expression"); // If I am understanding right
}
```

File: compiler/include/ForLoop.h [hunk 3]
```diff
@@ -132,6 +135,19 @@ class ForLoop final : public LoopStmt
...
+  bool needToHandleOuterVars() const override { return true; }
+  BlockStmt* loopBody() const override {
+    return const_cast<ForLoop*>(this);
+  }
+  bool needsInitialAccumulate() const override { return false; }
+  Expr* asExpr() override { return this; }
+  bool isInductionVar(Symbol* sym) override;
+
+  virtual bool isForallStmt() override { return false; }
+  virtual ForallStmt *forallStmt() override { INT_ASSERT(false); return nullptr; }
```
Comment (e-kayrakli) [inline compiler/include/ForLoop.h:149]: Nit, but for these two, consider:

- `final`
- dropping `virtual` (https://stackoverflow.com/questions/39932391/should-i-use-virtual-override-or-both-keywords)
- Not asserting for `forallStmt` and just returning `nullptr`.

File: compiler/include/ForallStmt.h [hunk 4]
```diff
@@ -69,24 +69,38 @@ class ForallOptimizationInfo {
     // forall loop statement //
 ///////////////////////////////////
 
-class ForallStmt final : public Stmt
+struct ShadowVarLoopInterface {
```
Comment (e-kayrakli) [inline compiler/include/ForallStmt.h:72]: Can we find a better place to live for this?

Also consider `LoopWithShadowVars` as a name alternative.

File: compiler/include/ForallStmt.h [hunk 5]
```diff
@@ -138,6 +152,11 @@ class ForallStmt final : public Stmt
 
   void insertZipSym(Symbol *sym);
 
+  bool isInductionVar(Symbol* sym) override;
+
+  virtual bool isForallStmt() override { return true; }
+  virtual ForallStmt *forallStmt() override { return this; }
```
Comment (e-kayrakli) [inline compiler/include/ForallStmt.h:158]: Ditto:

- `final`? At least for the first one
- no `virtual`

File: compiler/include/baseAST.h [hunk 6]
```diff
@@ -599,6 +599,7 @@ static inline const CallExpr* toConstCallExpr(const BaseAST* a)
       AST_CALL_CHILD(_a, WhileStmt,    condExprGet(),  call, __VA_ARGS__); \
                                                                            \
     } else if (isForLoop(_a)      == true) {                               \
+      AST_CALL_LIST (_a, ForLoop,      shadowVariables(), call, __VA_ARGS__); \
```
Comment (e-kayrakli) [inline compiler/include/baseAST.h:602]: Just noting that order of these AST* calls typically matter. There are some AST walks that start from some node and goes `->next` until it hits a marker etc. This order is a bit different than `forall`s. But you're not introducing that difference by this -- it seems to be there already.

File: compiler/passes/cleanup.cpp [hunk 7]
```diff
@@ -182,13 +182,19 @@ static void backPropagate(BaseAST* ast) {
 }
 
 static void handleNonTypedAndNonInitedVar(DefExpr* def) {
-  if (toVarSymbol(def->sym)) {
+  if (Symbol *sym = toVarSymbol(def->sym)) {
+    if(toShadowVarSymbol(sym)) {
```
Comment (e-kayrakli) [inline compiler/passes/cleanup.cpp:186]: Probably these two ifs can be combined into `if (isShadowVarSymbol(def->sym))`

File: compiler/passes/convert-uast.cpp [hunk 8]
```diff
@@ -1401,6 +1401,10 @@ struct Converter {
       } else if (const uast::ReduceIntent* rd = expr->toReduceIntent()) {
         astlocMarker markAstLoc(rd->id());
 
+        if(fForeachIntents && parent->toForeach()) {
```
Comment (e-kayrakli) [inline compiler/passes/convert-uast.cpp:1404]: Using `fForeachIntents` check here feels redundant, but I am still reading code.

File: compiler/passes/convert-uast.cpp [hunk 9]
```diff
@@ -1972,14 +1976,15 @@ struct Converter {
     // Does not appear possible right now, from reading the grammar.
     INT_ASSERT(!node->isExpressionLevel());
 
-    if (node->withClause()) {
+    if (!fForeachIntents && node->withClause()) {
```
Comment (e-kayrakli) [inline compiler/passes/convert-uast.cpp:1979]: This one feels more correct as this one's more fundamental.

File: compiler/passes/scopeResolve.cpp [hunk 10]
```diff
@@ -1474,7 +1475,17 @@ static void setupShadowVars() {
...
 
-  // Instead of the two nested loops above, we could march through
+  forv_Vec(BlockStmt, bs, gBlockStmts)
+    if(ForLoop *fl = toForLoop(bs)) {
+      for_shadow_vars_and_defs(svar, def, temp, fl) {
+        if (hasOuterVariable(svar))
+          setupOuterVar(fl, svar);
+        if (svar->isTaskPrivate())
+          checkRefsToIdxVars(fl, def, svar);
+      }
```
Comment (e-kayrakli) [inline compiler/passes/scopeResolve.cpp:1485]: This is just the second repetition between `forall` and `for`, but the core of the implementation is complex enough to deserve a helper.

File: compiler/resolution/functionResolution.cpp [hunk 11]
```diff
@@ -7661,7 +7661,7 @@ void printTaskOrForallConstErrorNote(Symbol* aVar) {
     Expr* enclLoop = aVar->defPoint->parentExpr;
 
     USR_PRINT(enclLoop,
-              "The shadow variable '%s' is constant due to forall intents "
+              "The shadow variable '%s' is constant due to intents "
```
Comment (e-kayrakli) [inline compiler/resolution/functionResolution.cpp:7664]: I wonder if we should consider "loop" intents. Pinging @bradcray to see if he has an opinion about this. I think we may be shifting towards being more open to using just "intent", but wanted to check.

File: compiler/resolution/functionResolution.cpp [hunk 12]
```diff
@@ -10596,9 +10596,23 @@ Expr* resolveExpr(Expr* expr) {
   } else if (SymExpr* se = toSymExpr(expr)) {
     makeRefType(se->symbol()->type);
 
-    if (ForallStmt* pfs = isForallIterExpr(se)) {
+    ForallStmt* pfs = isForallIterExpr(se);
+    ForLoop* pfl = toForLoop(se->parentExpr);
```
Comment (e-kayrakli) [inline compiler/resolution/functionResolution.cpp:10600]: Creating both pfs/pfl feels redundant again.

File: compiler/resolution/functionResolution.cpp (hunk 11 above)
Comment (bradcray) [reply compiler/resolution/functionResolution.cpp:7664]: Thinking about the general term for the concept, "Loop intents" doesn't feel right since the `begin` and `cobegin` cases aren't loops.  I still find "task intents" to be pretty good except for the concern that `foreach` loops don't technically introduce new tasks (unless we'd want to consider them "hardware tasks"?).  Of course, `for` loops don't introduce tasks either, but once they support intents it still seems like an accurate term in the sense of "the single task executing this loop will get this variable/intent."  "parallelism intents" is another possibility for the general term, though it feels a bit vaguer and like a mouthful.

In any case, I worry that simply saying "due to intents in this loop" feels vague without a modifier (even with the "in this loop" phrase that follows it).  And even the original wording gives me pause today because I wouldn't consider the compiler-introduced shadow variables as being an instance of a task intent (I don't think?).

So, reconsidering the message as a whole, I'd be tempted to say:

> '%s' is constant due to a compiler-introduced shadow variable at this point

for the case that the compiler introduced the shadow variable and 

> '%s' is constant due to the `with`-clause on this statement

when it doesn't.  Can we easily distinguish between those cases (by looking at a FLAG on `aVar`, say, like `FLAG_COMPILER_GENERATED` or `FLAG_TEMP`?).

File: compiler/AST/ForLoop.cpp [hunk 13]
```diff
@@ -297,6 +298,12 @@ BlockStmt* ForLoop::doBuildForLoop(Expr*      indices,
   loop->mContinueLabel = continueLabel;
   loop->mBreakLabel    = breakLabel;
 
+  // Transfer the DefExprs of the intent variables (ShadowVarSymbols).
+  if (isForeach && intents) {
```
[the remark this answers is not in the document]
Comment (stonea) [reply compiler/AST/ForLoop.cpp:302]: Sure, I can get rid of the `isForeach` part (at this point you'd still run into a syntax errors if you tried adding intents to a `for` loop).

File: compiler/AST/ForLoop.cpp (hunk 1 above)
Comment (stonea) [reply compiler/AST/ForLoop.cpp:660]: Yeah, if you did `foreach (i,j) in ..2D thing..` you'd see `mIndex->symbol` is a tuple (specifcally named `_indexOfInterest`).

Though looking at the AST it looks like we depack it into `i` and `j` and `i` and `j` get flagged as index variables.

So I'm not really sure if what I'm doing here is the right thing or not. But I tried running this example with a `forall` loop and it also only has that packed tuple on its list of induction variables, so that gives me some confidence.

Practically speaking I think what this is used for is to ensure we don't give an implicit intent to the loops index variable, so I'll try and keep this in mind when I start looking at implicit intents. I'm thinking maybe in the tuple case `i` and `j` are def'd in the loop body itself so that ends up not being a concern?

Anyway, let me know if you're comfortable with me leaving this "as is" for the moment or if you want me to dig deeper before merging.

File: compiler/include/ForallStmt.h (hunk 4 above)
Comment (stonea) [reply compiler/include/ForallStmt.h:72]: How about renaming it to LoopWithShadowVarsInterface? It's wordy but I like having 'Interface' in the name just so its obvious that this isn't a new node the AST or something like that.

And how about I make a new `LoopWithShadowVarsInterface.h` file for it? Alternatively I could put it in `ForLoop.h` or `intents.h` but neither of those seem like a good place to have it either.

File: compiler/include/baseAST.h (hunk 6 above)
Comment (stonea) [reply compiler/include/baseAST.h:602]: I want to say at some point I had this reordered but ran into use-before-def errors because of it.

File: compiler/passes/cleanup.cpp (hunk 7 above)
Comment (stonea) [reply compiler/passes/cleanup.cpp:186]: I can change it to this:

`if (isVarSymbol(def->sym) && !isShadowVarSymbol(def->sym))`

File: compiler/passes/convert-uast.cpp (hunk 8 above)
Comment (stonea) [reply compiler/passes/convert-uast.cpp:1404]: Yeah looks unnecessary.

I was a bit worried we might hit that error message before we check for the `with` clause more generally but I just tried it out and we don't. I'll simplify to just `if(parent->toForeach()) { ... }`

File: compiler/passes/scopeResolve.cpp (hunk 10 above)
Comment (stonea) [reply compiler/passes/scopeResolve.cpp:1485]: I assume you're suggesting that I factor out the common body of these two loops and call to that.

Sounds good, I refactored it to this:

```C++
  template<class T>
  static void setupShadowVarsOnLoop(T *loop) {
    INT_ASSERT(isForallStmt(loop) || isForLoop(loop));
    for_shadow_vars_and_defs(svar, def, temp, loop) {
      if (hasOuterVariable(svar))
        setupOuterVar(loop, svar);
       if (svar->isTaskPrivate())
        checkRefsToIdxVars(loop, def, svar);
    }
  }

  static void setupShadowVars() {
    forv_Vec(ForallStmt, fs, gForallStmts)
      setupShadowVarsOnLoop(fs);

    forv_Vec(BlockStmt, bs, gBlockStmts)
      if(ForLoop *fl = toForLoop(bs))
        setupShadowVarsOnLoop(fl);

    // Instead of the nested loops above, we could march through
    // gShadowVarSymbols and invoke setupOuterVar(svar->parentExpr, svar).
    // The nested loops group together all shadow variables of a given
    // ForallStmt, so hopefully have better cache behavior.

    USR_STOP();
  }
````

I guess ideally I could have `setupShadowVarsOnLoop` take a pointer to my new interface and use that but that doesn't seem to play well with `for_shadow_vars_and_defs` which is assuming a specifically named data member (which maybe I should change how that iterator works but I'll wait to see if this comes up a lot more).

File: compiler/resolution/functionResolution.cpp (hunk 12 above)
Comment (stonea) [reply compiler/resolution/functionResolution.cpp:10600]: In this case we really do, do different things depending on if its a foreach or forall loop. For forall loops we call `resolveForallHeader` which does some extra stuff in addition to `setupAndResolveShadowVars` (which we also do for `foreach` loops).

In the long run I might end up abstracting `resolveForallHeader` to work with `foreach` loops as well but I won't tackle that this PR.

So I think we still need both branches.

As far as doing something like `else if(ForLoop* pfl = ...) { ... }` instead of calculating `pfl` above, in this case since there's more to that condition I want to `&&` together that needs that pointer `pfl` to be assigned ahead of time. I could duplicate `toForLoop(se->parentExpr)` multiple times (uggh) or I could nest it but this is part of a larger chain of "if/else if" blocks so that's not easily doable either.

I'm going to factor this logic out into its own function and then call that (so atleast it looks less ugly at this site).

File: compiler/resolution/functionResolution.cpp (hunk 11 above)
Comment (stonea) [reply compiler/resolution/functionResolution.cpp:7664]: > Can we easily distinguish between those cases (by looking at a FLAG on aVar, say, like FLAG_COMPILER_GENERATED or FLAG_TEMP?)

Running it through the debugger and looking at `viewFlags` on `aVar` I don't see any flag that would let us disambiguate if the variable was coming from a shadow an example where the intent is implicit there's nothing that makes it obvious.

Not to say it can't be done, but I'm inclined just to reword this to say "is constant due to task intents in this loop" and move on (even though as you mentioned 'task' may not be the right word for `foreach` and eventually `for`).

I've spawned a new GH issue for follow up: https://github.com/chapel-lang/chapel/issues/23783

Though unless I get pushback, I'm not really planning to pick this up myself unless I see a way to easily fix it as I continue working on implementing `foreach` intents.  (and presumably this'll whole section of code will be subsumed by Dyno anyway).
